### ISSUE TITLE
manipulation of DDs

### ISSUE THREAD

**Issue by alexandrupaler** [iic-jku/dd_package#1]
Is there a way to view/manipulate the diagrams after these were generated from a circuit?

**Comment by hillmich** [comment]
In the Package class, there is a method `void toDot(Edge e, std::ostream& oss, bool isVector = false);` to export the decision diagram to the GraphViz format. To read and generate the DD, you should have a look an the [QFR](https://github.com/iic-jku/qfr).

What kind of manipulations do you have in mind?

**Comment by alexandrupaler** [comment]
Thank you that answers my question. I wrote the comment from the phone, and github does not render the readme.md with the instructions.

I am interested in manually changing the variable ordering and observing the resulting DD.

**Comment by hillmich** [comment]
Reordering a decision diagram in the package is currently not supported :/

**Comment by alexandrupaler** [comment]
Ok. Is it possible to specify a variable ordering that stays fixed during DD usage?

**Comment by hillmich** [comment]
There is no option to explicitly specify the variable ordering.
You may change the ordering of qubits in your circuit as a workaround.

**Comment by alexandrupaler** [comment]
I see. Is qubit ordering based on the .real file header?

**Comment by hillmich** [comment]
Yeah, that should work (we never tested this though :sweat_smile: ).

The first variable in real header will be on the bottom of the DD, the last variable will be the root.

**Comment by alexandrupaler** [comment]
Can you please point to studies relating variable ordering and qmdd dimensions?

**Comment by hillmich** [comment]
Some older works: 
https://ieeexplore.ieee.org/abstract/document/4215933 
https://pdfs.semanticscholar.org/ffa5/e583fd842679f377ab4c8e381205aec51f5b.pdf

I'm not aware of anything more recent.

Regarding your second question, that is certainly possible but not implemented in such a way as of now.

**Comment by alexandrupaler** [comment]
Ok. So the only way for reordering is to change the .real header. What happens in the case of qasm or grcs circuits?

**Comment by hillmich** [comment]
If you have a register for each individual qubit you could change the order of your `qreg ...` statements. There is no way to achieve this for the GRCS files.

**Comment by alexandrupaler** [comment]
I ran a kind of own-developed analysis on a particular quantum circuit. Tested two qubit orderings, and the results wrt to number of nodes in the diagram are below:
```
Config1 min 442 max 598 avg 515.15234375
Config2 min 586 max 658 avg 629.453125
```
It can be seen that for this particular quantum circuit, which was not randomly generated, about 20% savings can be achieved by reordering.

**Comment by hillmich** [comment]
Are you at the institute anytime soon to discuss this?

**Comment by alexandrupaler** [comment]
We can discuss over email. I am in Texas for the following months.
